### ISSUE TITLE
Definition of octal codes in literal strings related to UTF-16BE encoding

### ISSUE THREAD

**Issue by jmlehton** [pdf-association/pdf-issues#494]
We feel that there is a small unclarity regarding to literal strings in ISO 32000-2:2020 (and previous versions). In ch. 7.3.4.2 "Literal strings" Table 3, a single "\ddd" octal code is defined as a "character code". Isn't a "character code" something which maps to a character in a codepage in question? For strings encoded with UTF-16BE, a single octal code can not really be used as a mapping character code (i.e. \ddd does not map to a Unicode character). Of course this can be done with multiple octal codes, but the definition is about a single octal code \ddd. From this, it may be unclear for the reader whether it is possible to use octal coding to UTF-16BE encoded string with multiple-byte characters.

It is true that ch. 7.3.4.2 "Literal strings" also states that any 8-bit value can appear also with the octal "notation described". But this still can be understood so that "notation described" refers to defining (i.e. limiting) an octal code as a mapping character code, which leads to the original unclarity.

In future revisions, we suggest to reconsider or open the term "character code" used in octal codes and to give a short sentence about its usage with Unicode in a case where a single character requires multiple bytes.

**Comment by petervwyatt** [comment]
I cannot find the use of the phrase "_character code_" anywhere in 7.3.4 subclauses - and that would be incorrect. But I agree that the language is confusing since the word "_character_" is used for both the bytes comprising the string in the input PDF as well as what they mean once lexed/de-escaped:

7.3.4.1: "_A string object shall consist of a series of zero or more bytes._"
7.3.4.2: "_A literal string shall be written as an arbitrary number of characters enclosed ..._"

The correct terminology should be "_characters_" are what comprise the string in "raw PDF" (pre-lexing), but "_bytes_" are what they represent post-lexing. So an octal code `\ddd` in a literal string comprises up to 4 "characters" but presents a single "byte" in that string object. The interpretation of those bytes (such as in a specific encoding) is dependent on type definitions elsewhere in the spec, and according to 7.9.2 and Figure 7.

**Comment by petervwyatt** [comment]
Proposed solution:
> The `\ddd` escape sequence provides a way to represent **<del>characters</del>** _<ins>bytes</ins>_ outside the printable ASCII character set.

and

> Since any 8-bit value may appear in a string (with proper escaping for REVERSE SOLIDUS (backslash) and unbalanced PARENTHESES) this `\ddd` notation provides a way to specify **<del>characters</del>** _<ins>bytes</ins>_ outside the ASCII character set by using ASCII characters only. However, any 8-bit value may appear in a string, represented either as itself or with the \ddd notation described.

**Comment by petervwyatt** [comment]
I agree 7.3.4.1, 1st bullet should drop the word "literal" - it should just state "characters" so it is consistent with the terminology throughout 7.3.4.2:

> * As a sequence of <del>literal</del> characters enclosed in parentheses () (using LEFT PARENTHESIS (28h) and RIGHT PARENTHESIS (29h)); see 7.3.4.2, "Literal strings"

I think other errata we have already applied sufficiently cover encodings and the fact that the lexical form of a string object is orthogonal to any character encoding in string data - see from this point down https://pdf-issues.pdfa.org/32000-2-2020/clause07.html#H7.9.1

**Comment by jmlehton** [comment]
> I cannot find the use of the phrase "_character code_" anywhere in 7.3.4 subclauses - and that would be incorrect.

The "character code" definition is in the last row of Table 3 (in ch.  7.3.4.2).

**Comment by car222222** [comment]
That definition could be made more pre use, and understandable as follow:

\ddd     ASCII Character code ddd (3 octal digits)

(Assuming I interpreted it correctly 😄 !)

**Comment by car222222** [comment]
I edited this last comment to make the following correction:
  ASCII  changed to  8-bit

**Comment by petervwyatt** [comment]
Thanks. Table 3 proposed fix is quite simple - its a byte:

> `\ddd`   Byte with value `ddd` in octal 

I would NOT add "(3 octal digits)" as that is incorrect - it can be 1-3 bytes as per normative text further down.

**Comment by car222222** [comment]
OK with me.

**Comment by jmlehton** [comment]
> Thanks. Table 3 proposed fix is quite simple - its a byte:
> 
> > `\ddd`   Byte with value `ddd` in octal
> 
> I would NOT add "(3 octal digits)" as that is incorrect - it can be 1-3 bytes as per normative text further down.

This is great. Thanks. And yes, "(3 octal digits)" would be incorrect.

**Comment by petervwyatt** [comment]
PDF TWG agree
